File: test/perf/micro/scala/src/main/scala/perf.scala
// This file is a part of Julia. License is MIT: http://julialang.org/license

import scala.util._
import java.io._
import breeze.linalg._
import breeze.numerics._
import breeze.stats._
import breeze.math._
//import com.github.fommil.netlib.{BLAS}

object PerfBreeze {
  final val NITER = 5

  // print results appropriately. times are in milliseconds
  def print_perf(name:String, t:Double) = {
    printf("scala,%s,%.9f\n", name, t/1e6)
  }

  // time fib
  def fib(n:Int):Int = {
    if (n < 2) n else fib(n-1) + fib(n-2)
  }

  def time_fib() = {
    assert(fib(20) == 6765)
    var tmin = Long.MaxValue
    var f = 0

    for(i <- 1 to NITER) {
      val t1 = System.nanoTime()
      for(j <- 1 to 1000) {
        f += fib(20)
      }
      val t = System.nanoTime() - t1
      if(t < tmin) tmin = t
    }

    tmin / 1000.0
  }

  // time parseint
  def time_parseint() = {
    val generator = scala.util.Random
    var tmin = Long.MaxValue

    for(i <- 1 to NITER) {
      var rand:Int = 0
      var rands:String = "0"
      var parsed:Int = 0
      val t1 = System.nanoTime()
      for(j <- 1 to 1000) {
        rand = generator.nextInt()
        rands = if(rand < 0) "-" + abs(rand).toHexString else rand.toHexString
        parsed = Integer.parseInt(rands, 16)
      }
      val t = System.nanoTime() - t1
      assert(rand == parsed)
      if(t < tmin) tmin = t
    }
    tmin / 1000.0
  }

  // time mandel
  def mandel(zin:Complex):Int = {
    val c = zin
    var z = zin
    val maxiter = 80
    for(n <- 0 to maxiter) {
      if(z.abs > 2) return n
      z = c + (z * z)
    }
    maxiter
  }

  def mandelperf() = {
    for(re <- -20 to 5; im <- -10 to 10) yield mandel(re/10.0 + i * im/10.0)
  }

  def time_mandel() = {
    var mandel_sum = 0
    var mandel_sum2 = 0
    var tmin = Long.MaxValue

    for(i <- 1 to NITER) {
      val t1 = System.nanoTime()
      for(j <- 1 to 100) {
        val mandel_arr = mandelperf()
        if(j == 1) {
          mandel_sum = sum(mandel_arr)
          mandel_sum2 += mandel_sum
        }
      }
      val t = System.nanoTime() - t1
      if(t < tmin) tmin = t
    }
    assert(mandel_sum == 14791)
    assert(mandel_sum2 == mandel_sum * NITER)
    tmin / 100.0
  }

  // time quicksort
  def quicksort(a:Array[Double], lo:Int, hi:Int):Array[Double] = {
    var i, l = lo
    var j = hi

    def _swap(i:Int, j:Int) = {
      val tmp = a(i)
      a(i) = a(j)
      a(j) = tmp
    }

    while(i < hi) {
      val pivot = a((l+hi)>>>1)
      while(i <= j) {
        while(a(i) < pivot) i += 1
        while(a(j) > pivot) j -= 1
        if(i <= j) {
          _swap(i, j)
          i += 1
          j -= 1
        }
      }
      if(l < j) quicksort(a, l, j)
      l = j
      j = hi
    }
    a
  }

  /*
  def checksorted(a:Array[Double]):Boolean = {
    for(i <- 0 to a.length-2) {
      assert(a(i) < a(i+1))
    }
    true
  }
  */

  def time_quicksort() = {
    var tmin = Long.MaxValue

    for(i <- 1 to NITER) {
      val t1 = System.nanoTime()
      for(j <- 1 to 1000) {
        val A = DenseVector.rand[Double](5000)
        quicksort(A.data, 0, 4999)
      }
      val t = System.nanoTime() - t1
      if(t < tmin) tmin = t
    }
    tmin / 1000.0
  }

  // time pisum
  def pisum() = {
    var sum = 0.0
    for(j <- 1 to 500) {
      sum = 0.0
      for(k <- 1 to 10000) {
        sum += 1.0/(k*k)
      }
    }
    sum
  }

  def time_pisum() = {
    var tmin = Long.MaxValue
    var pi = 0:Double
    for(i <- 1 to NITER) {
      val t1 = System.nanoTime()
      pi = pisum()
      val t = System.nanoTime() - t1
      if(t < tmin) tmin = t
      assert(abs(pi-1.644834071848065) < 1e-12)
    }
    tmin
  }

  // time printfd
  def printfd(n:Int) = {
    var stream = None: Option[PrintStream]
    try {
      stream = Some(new PrintStream(new BufferedOutputStream(new FileOutputStream("/dev/null"))))
      val valid_stream = stream.get
      for (i <- 1 to n) {
        valid_stream.printf(i + " " + i)
      }
    } catch {
      case e: Exception => println("Exception caught: " + e)
    } finally {
      if(stream.isDefined) stream.get.close()
    }
  }

  def time_printfd() = {
    var tmin = Long.MaxValue
    for(i <- 1 to NITER) {
      val t1 = System.nanoTime()
      printfd(100000)
      val t = System.nanoTime() - t1
      if(t < tmin) tmin = t
    }
    tmin
  }

  // random matrix statistics
  def randmatstat(t:Int):(Double,Double) = {
    val n = 5
    val v = DenseVector.zeros[Double](t)
    val w = DenseVector.zeros[Double](t)

    val g = breeze.stats.distributions.Gaussian(0, 1)
    for(i <- 0 to t-1) {
      val a = DenseMatrix.rand(n, n, g)
      val b = DenseMatrix.rand(n, n, g)
      val c = DenseMatrix.rand(n, n, g)
      val d = DenseMatrix.rand(n, n, g)
      val P = DenseMatrix.horzcat(a, b, c, d)
      val Q = DenseMatrix.vertcat(DenseMatrix.horzcat(a, b), DenseMatrix.horzcat(c, d))
      val V = P.t * P
      val W = Q.t * Q

      v(i) = trace(V * V * V * V)
      w(i) = trace(W * W * W * W)
    }
    (stddev(v)/mean(v), stddev(w)/mean(w))
  }

  def time_randmatstat() = {
    var tmin = Long.MaxValue
    for(i <- 1 to NITER) {
      val t1 = System.nanoTime()
      val (s1, s2) = randmatstat(1000)
      val t = System.nanoTime() - t1
      assert(0.5 < s1 && s1 < 1.0 && 0.5 < s2 && s2 < 1.0)

      if(t < tmin) tmin = t
    }
    tmin
  }

  // random matrix multiplication
  def randmatmul(t:Int):DenseMatrix[Double] = {
    val m1 = randomDouble((t, t))
    val m2 = randomDouble((t, t))
    m1 * m2
  }

  def time_randmatmul() = {
    var tmin = Long.MaxValue
    for(i <- 1 to NITER) {
      val t1 = System.nanoTime()
      val m = randmatmul(1000)
      val t = System.nanoTime() - t1
      assert(0 <= m(0,0))

      if(t < tmin) tmin = t
    }
    tmin
  }


  def main(args: Array[String]) = {
    //println("BLAS: " + BLAS.getInstance().getClass().getName())
    print_perf("fib", time_fib())
    print_perf("parse_int", time_parseint())
    print_perf("mandel", time_mandel())
    print_perf("quicksort", time_quicksort())
    print_perf("pi_sum", time_pisum())
    print_perf("rand_mat_stat", time_randmatstat())
    print_perf("rand_mat_mul", time_randmatmul())
    print_perf("printfd", time_printfd())
  }
}

File: test/perf/micro/perf.py
from numpy import *
from numpy.random import rand, randn
from numpy.linalg import matrix_power
import sys
import time
import random

if sys.version_info < (3,):
    range = xrange

## fibonacci ##

def fib(n):
    if n<2:
        return n
    return fib(n-1)+fib(n-2)

## quicksort ##

def qsort_kernel(a, lo, hi):
    i = lo
    j = hi
    while i < hi:
        pivot = a[(lo+hi) // 2]
        while i <= j:
            while a[i] < pivot:
                i += 1
            while a[j] > pivot:
                j -= 1
            if i <= j:
                a[i], a[j] = a[j], a[i]
                i += 1
                j -= 1
        if lo < j:
            qsort_kernel(a, lo, j)
        lo = i
        j = hi
    return a

## randmatstat ##

def randmatstat(t):
    n = 5
    v = zeros(t)
    w = zeros(t)
    for i in range(t):
        a = randn(n, n)
        b = randn(n, n)
        c = randn(n, n)
        d = randn(n, n)
        P = concatenate((a, b, c, d), axis=1)
        Q = concatenate((concatenate((a, b), axis=1), concatenate((c, d), axis=1)), axis=0)
        v[i] = trace(matrix_power(dot(P.T,P), 4))
        w[i] = trace(matrix_power(dot(Q.T,Q), 4))
    return (std(v)/mean(v), std(w)/mean(w))

## randmatmul ##

def randmatmul(n):
    A = rand(n,n)
    B = rand(n,n)
    return dot(A,B)

## mandelbrot ##

def abs2(z):
    return real(z)*real(z) +  imag(z)*imag(z)

def mandel(z):
    maxiter = 80
    c = z
    for n in range(maxiter):
        if abs2(z) > 4:
            return n
        z = z*z + c
    return maxiter

def mandelperf():
    r1 = [-2. + 0.1*i for i in range(26)]
    r2 = [-1. + 0.1*i for i in range(21)]
    return [mandel(complex(r, i)) for r in r1 for i in r2]

def pisum():
    sum = 0.0
    for j in range(1, 501):
        sum = 0.0
        for k in range(1, 10001):
            sum += 1.0/(k*k)
    return sum

#### Is this single threaded?
# def pisumvec():
#     return numpy.sum(1./(numpy.arange(1,10000)**2))

def parse_int(t):
    for i in range(1,t):
        n = random.randint(0,2**32-1)
        s = hex(n)
        if s[-1]=='L':
            s = s[0:-1]
        m = int(s,16)
        assert m == n
    return n

def printfd(t):
    f = open("/dev/null", "w")
    for i in range(1,t):
        f.write("{:d} {:d}\n".format(i, i+1))
    f.close()
    

def print_perf(name, time):
    print("python," + name + "," + str(time*1000))

## run tests ##

if __name__=="__main__":

    mintrials = 5

    assert fib(20) == 6765
    tmin = float('inf')
    for i in range(mintrials):
        t = time.time()
        f = fib(20)
        t = time.time()-t
        if t < tmin: tmin = t
    print_perf("recursion_fibonacci", tmin)

    tmin = float('inf')
    for i in range(mintrials):
        t = time.time()
        n = parse_int(1000)
        t = time.time()-t
        if t < tmin: tmin = t
    print_perf ("parse_integers", tmin)

    assert sum(mandelperf()) == 14791
    tmin = float('inf')
    for i in range(mintrials):
        t = time.time()
        mandelperf()
        t = time.time()-t
        if t < tmin: tmin = t
    print_perf ("userfunc_mandelbrot", tmin)

    tmin = float('inf')
    for i in range(mintrials):
        lst = [ random.random() for i in range(1,5000) ]
        t = time.time()
        qsort_kernel(lst, 0, len(lst)-1)
        t = time.time()-t
        if t < tmin: tmin = t
    print_perf ("recursion_quicksort", tmin)

    assert abs(pisum()-1.644834071848065) < 1e-6
    tmin = float('inf')
    for i in range(mintrials):
        t = time.time()
        pisum()
        t = time.time()-t
        if t < tmin: tmin = t
    print_perf ("iteration_pi_sum", tmin)

    # assert abs(pisumvec()-1.644834071848065) < 1e-6
    # tmin = float('inf')
    # for i in range(mintrials):
    #     t = time.time()
    #     pisumvec()
    #     t = time.time()-t
    #     if t < tmin: tmin = t
    # print_perf ("pi_sum_vec", tmin)

    (s1, s2) = randmatstat(1000)
    assert s1 > 0.5 and s1 < 1.0
    tmin = float('inf')
    for i in range(mintrials):
        t = time.time()
        randmatstat(1000)
        t = time.time()-t
        if t < tmin: tmin = t
    print_perf ("matrix_statistics", tmin)

    tmin = float('inf')
    for i in range(mintrials):
        t = time.time()
        C = randmatmul(1000)
        assert C[0,0] >= 0
        t = time.time()-t
        if t < tmin: tmin = t
    print_perf ("matrix_multiply", tmin)

    tmin = float('inf')
    for i in range(mintrials):
        t = time.time()
        printfd(100000)
        t = time.time()-t
        if t < tmin: tmin = t
    print_perf ("print_to_file", tmin)
